lib/recursive-methods.rb

```diff
@@ -1,49 +1,105 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the value of n
+# Space complexity: O(n) where n is the value of n
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif n <= 1
+    return 1
+  else 
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s 
+  else 
+    return s[s.length - 1] + reverse(s[0..-2])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s 
+  else 
+    return s[s.length - 1] + reverse(s[0..-2])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is value of n
+# Space complexity: O(n) where n is value of n
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 1
+    return 0
+  else 
+    return 2 + bunny(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return true
+  else 
+    if s[0] == "(" && s[-1] == ")"
+      return nested(s[1..-2])
+    else
+      return false
+    end
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of array
+# Space complexity: O(n) where n is the length of array
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  if array.empty?
+    return false
+  else
+    if array[0] == value
+      return true
+    else
+      array.shift
```

Comment:
This shifts all the elements in the array over one index (this one line is O(n)).